include/storage/shared_datatype.hpp

```diff
@@ -160,40 +160,111 @@ class DataLayout
 };
 
-enum SharedDataType
+struct SharedRegion
 {
-    REGION_NONE,
-    REGION_1,
-    REGION_2
-};
+    static constexpr const int MAX_NAME_LENGTH = 254;
 
-struct SharedDataTimestamp
-{
-    explicit SharedDataTimestamp(SharedDataType region, unsigned timestamp)
-        : region(region), timestamp(timestamp)
+    SharedRegion() : name{0}, timestamp{0} {}
+    SharedRegion(const std::string &name_, std::uint64_t timestamp, std::uint8_t shm_key)
+        : name{0}, timestamp{timestamp}, shm_key{shm_key}
     {
+        std::copy_n(name_.begin(), std::min<std::size_t>(MAX_NAME_LENGTH, name_.size()), name);
     }
 
-    SharedDataType region;
-    unsigned timestamp;
+    bool IsEmpty() const { return timestamp == 0; }
 
-    static constexpr const char *name = "osrm-region";
+    char name[MAX_NAME_LENGTH + 1];
+    std::uint64_t timestamp;
+    std::uint8_t shm_key;
 };
 
-inline std::string regionToString(const SharedDataType region)
+// Keeps a list of all shared regions in a fixed-sized struct
+// for fast access and deserialization.
+struct SharedRegionRegister
 {
-    switch (region)
+    using RegionID = std::uint8_t;
+    static constexpr const RegionID INVALID_REGION_ID = std::numeric_limits<RegionID>::max();
+    using ShmKey = decltype(SharedRegion::shm_key);
+
+    // Returns the key of the region with the given name
+    RegionID Find(const std::string &name) const
     {
-    case REGION_1:
-        return "REGION_1";
-    case REGION_2:
-        return "REGION_2";
-    case REGION_NONE:
-        return "REGION_NONE";
-    default:
-        return "INVALID_REGION";
+        auto iter = std::find_if(regions.begin(), regions.end(), [&](const auto &region) {
+            return std::strncmp(region.name, name.c_str(), SharedRegion::MAX_NAME_LENGTH) == 0;
+        });
+
+        if (iter == regions.end())
+        {
+            return INVALID_REGION_ID;
+        }
+        else
+        {
+            return std::distance(regions.begin(), iter);
+        }
     }
-}
+
+    RegionID Register(const std::string &name, ShmKey key)
+    {
+        auto iter = std::find_if(
+            regions.begin(), regions.end(), [&](const auto &region) { return region.IsEmpty(); });
+        if (iter == regions.end())
+        {
+            throw util::exception("No shared memory regions left. Could not register " + name +
+                                  ".");
+        }
+        else
+        {
+            constexpr std::uint32_t INITIAL_TIMESTAMP = 1;
+            *iter = SharedRegion{name, INITIAL_TIMESTAMP, key};
+            RegionID key = std::distance(regions.begin(), iter);
+            return key;
+        }
+    }
+
+    template <typename OutIter> void List(OutIter out) const
+    {
+        for (const auto &region : regions)
+        {
+            if (!region.IsEmpty())
+            {
+                *out++ = region.name;
+            }
+        }
+    }
+
+    void Deregister(const RegionID key) { regions[key] = SharedRegion{}; }
```

Comment:
The method is not used and `DataWatchdogImpl` assumes that the shared region pointer won't be changed during an engine process life-time. Is a follow-up issue about deleting named regions needed?

Reply:
Good catch, I wanted to use this for the spring-clean command, but instead we just overwrite the whole registry.